### PR TITLE
Add initial bert example

### PR DESCRIPTION
**Pull request by mattdangerw**
This adds an initial BERT code example which we can use to test our upcoming
components (transformer encoder, positional embedding, tokenizers) in
an end to end BERT model.

This uses code from the official bert repository and model garden.
It attempts to be complete--it contains scripts for sentence splitting raw
input data with nltk and computing a wordpiece vocab with tf_text.

Still to come:

 - AdamW optimizer
 - Proper learning rate schedule
 - TPU strategy support
 - GLUE finetuning script

### CODE REVIEW

**Comment by mattdangerw** [comment]
Happy to split things up into separate PRs, which might be easier for reviewing, but the advantage of one big one is the whole set can be tested out. I've included some small amount of example data to exercise the script.

Running the following from the base of the repo should take raw wiki text input, split it, create pretraining data with masks and next sentence labels, and train a tiny and undertrained BERT model.

```
DIR=./bert_test_output
mkdir $DIR
python3 -m venv $DIR/env
source $DIR/env/bin/activate
pip install -e ".[tests,examples]"
python3 examples/bert/create_sentence_split_data.py \
    --input_files examples/bert/data/wiki_example_data.txt \
    --output_directory $DIR/sentences --num_shards 1
python3 examples/bert/create_pretraining_data.py \
    --input_files $DIR/sentences/ \
    --vocab_file examples/bert/data/vocab.txt \
    --output_file $DIR/data/pretraining.tfrecord
python3 examples/bert/run_pretraining.py \
    --input_files $DIR/data/ \
    --vocab_file examples/bert/data/vocab.txt \
    --bert_config_file examples/bert/data/bert_tiny.json \
    --saved_model_output $DIR/model/
```

File: examples/bert/bert_model.py [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+
+    def call(self, inputs):
+        """Transformer self-attention encoder block call.
+
+        Args:
+            inputs: a single tensor or a list of tensors.
+                `input tensor` as the single sequence of embeddings.
+                [`input tensor`, `attention mask`] to have the additional
+                attention mask.
+                [`query tensor`, `key value tensor`, `attention mask`] to have
```
Comment (fchollet) [inline examples/bert/bert_model.py:305]: Rather than nested arguments, we should prefer explicit keyword arguments.

We should also make the layer compatible with the built-in masking system. This means that if the query or key have masked values, we should take them into account in the attention mask. Let's discuss this

File: examples/bert/bert_model.py [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+        if self._norm_first:
+            return source_attention_output + layer_output
+
+        # During mixed precision training, layer norm output is always fp32 for
+        # now. Casts fp32 for the subsequent add.
+        layer_output = tf.cast(layer_output, tf.float32)
+        return self._output_layer_norm(layer_output + attention_output)
+
+
+class PositionEmbedding(keras.layers.Layer):
```
Comment (fchollet) [inline examples/bert/bert_model.py:373]: Do we ever need position embeddings in a standalone way, or should we make this layer the combination of a regular embedding layer and a position embedding layer? This would be much simpler for end users

File: examples/bert/bert_model.py [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+        )
+
+        super().build(input_shape)
+
+    def call(self, inputs):
+        flat_inputs = tf.reshape(inputs, [-1])
+        if self._use_one_hot:
+            dtype = self._compute_dtype
+            if not tf.dtypes.as_dtype(dtype).is_floating:
+                # TensorFlow 1 compatibility. In TF1, self._compute_dtype is
```
Comment (fchollet) [inline examples/bert/bert_model.py:501]: We can discard all TF1 compat concerns, this is a TF2 only codebase.

File: examples/bert/bert_model.py [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+
+class OnDeviceEmbedding(keras.layers.Layer):
+    """Performs an embedding lookup suitable for accelerator devices.
+
+    This layer uses either tf.gather or tf.one_hot to translate integer indices
+    to float embeddings.
+
+    Args:
+        vocab_size: Number of elements in the vocabulary.
+        embedding_width: Output size of the embedding layer.
```
Comment (fchollet) [inline examples/bert/bert_model.py:447]: We should make the API consistent with the `Embedding` layer.

File: examples/bert/bert_model.py [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+        actual_seq_len = input_shape[self._seq_axis]
+        position_embeddings = self._position_embeddings[:actual_seq_len, :]
+        new_shape = [1 for _ in inputs.get_shape().as_list()]
+        new_shape[self._seq_axis] = actual_seq_len
+        new_shape[-1] = position_embeddings.get_shape().as_list()[-1]
+        position_embeddings = tf.reshape(position_embeddings, new_shape)
+        return tf.broadcast_to(position_embeddings, input_shape)
+
+
+class OnDeviceEmbedding(keras.layers.Layer):
```
Comment (fchollet) [inline examples/bert/bert_model.py:439]: This layer doesn't actually do device placement. It seems to require quite a bit of knowledge of the underlying implementation in order to be used correctly. If possible, I recommend make it an implementation detail of the regular embedding / positional embedding layer.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Some preliminary comments. I'm reviewing the layers with the assumption that we want them to be reusable and part of the API.

File: examples/bert/utils.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2022 The KerasNLP Authors
```
Comment (mattdangerw) [inline examples/bert/utils.py:1]: Rename this bert_utils.py (for search)

File: examples/bert/run_pretraining.py [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+        )
+        output_tensor = tf.gather(flat_sequence_tensor, flat_positions)
+
+        return output_tensor
+
+
+class BertPretrainer(keras.Model):
+    def __init__(self, bert_model, **kwargs):
+        super().__init__(**kwargs)
+        self.bert_model = bert_model
```
Comment (mattdangerw) [inline examples/bert/run_pretraining.py:236]: create metrics here

File: examples/__init__.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2021 The KerasNLP Authors
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/__init__.py:1]: done

File: examples/bert/bert_model.py [hunk 9]
```diff
@@ -0,0 +1,709 @@
...
+
+    inputs[0]: from_tensor: 2D or 3D Tensor of shape
+        [batch_size, from_seq_length, ...].
+    inputs[1]: to_mask: int32 Tensor of shape [batch_size, to_seq_length].
+
+    Returns:
+        float Tensor of shape [batch_size, from_seq_length, to_seq_length].
+    """
+
+    def call(self, inputs, to_mask=None):
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/bert_model.py:34]: yeah, bad name, this is the 2d, input mask, renamed

File: examples/bert/bert_model.py [hunk 10]
```diff
@@ -0,0 +1,709 @@
...
+"""Bert model and layer implementations.
+
+We should work to replace this with components from the keras-nlp library.
+"""
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+class SelfAttentionMask(keras.layers.Layer):
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/bert_model.py:23]: utility function def better, I switched it for a utility function

File: examples/bert/bert_model.py (hunk 1 above)
Comment (mattdangerw) [reply examples/bert/bert_model.py:305]: Yeah, added named arguments, which helps. Using built in masking I left as a TODO, we can clean that up as we make this an actual library layer. For now it takes an `attention_mask` kwarg, which is equivalent to `MultiHeadAttention` in core keras.

File: examples/bert/bert_model.py (hunk 2 above)
Comment (mattdangerw) [reply examples/bert/bert_model.py:373]: I am in favor of keeping things modular I think. We might sub out this embedding for a sinusoidal embedding, and might have already performed the term embedding lookup elsewhere.

I think things get less flexible if you fuse, and at least for bert you haven't really the complexity of summing multiple embeddings (still need type embedding for next sentence prediction).

This would be a good discussion to have as we build PositionEmbedding into the library.

File: examples/bert/bert_model.py (hunk 5 above)
Comment (mattdangerw) [reply examples/bert/bert_model.py:439]: Yeah, this is strictly for TPU support IIUC. We should remove it entirely and make any changes to core embedding layer to do so. Left a TODO.

File: examples/bert/bert_model.py (hunk 4 above)
Comment (mattdangerw) [reply examples/bert/bert_model.py:447]: I somewhat think vocab_size is a nicer name that input_dim given that the input shape need have nothing to do with input_dim. But we can't really change keras.layers.Embedding, so could rename to input_dim, and output_dim everywhere. wdyt?

File: examples/bert/create_pretraining_data.py [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+    masked_lm_positions = []
+    masked_lm_labels = []
+    for p in masked_lms:
+        masked_lm_positions.append(p.index)
+        masked_lm_labels.append(p.label)
+
+    return (output_tokens, masked_lm_positions, masked_lm_labels)
+
+
+def truncate_seq_pair(tokens_a, tokens_b, max_num_tokens, rng):
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/create_pretraining_data.py:458]: Would rather do this as a follow up. Let's start with the stock "create_pretraining_data" from the initial bert and tweaks as desired later.

File: examples/bert/create_pretraining_data.py [hunk 12]
```diff
@@ -0,0 +1,524 @@
...
+                    vocab,
+                    rng,
+                )
+            )
+    rng.shuffle(instances)
+    return instances
+
+
+def create_instances_from_document(
+    all_documents,
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/create_pretraining_data.py:251]: No I don't think so, all_documents is used for the next sentence prediction (sometime you need to pull a random sentence from a different document) if I remember correctly.

File: examples/bert/README.md [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+Preparing the pretraining data will happen in two stages. First, raw text needs
+to be split into lists of sentences per document. Second, this sentence split
+data needs to use to create training examples with both masked words and
+next sentence predictions.
+
+### Splitting raw text into sentences
+
+The `create_sentence_split_data.py` will process raw input files and split them
+into output files where each line contains a sentence, and a blank line marks
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/README.md:89]: good call. done!

File: examples/bert/README.md [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+English uncased vocabulary
+[here](https://storage.googleapis.com/tensorflow/keras-nlp/examples/bert/bert_vocab_uncased.txt).
+
+### Tokenize, mask, and combine sentences into training examples
+
+The `create_pretraining_data.py` scrip will take in a set of sentence split
+files, and set up training examples for the next sentence prediction and masked
+word tasks.
+
+The output of the script will be TFRecord files with a number of fields per
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/README.md:121]: Added one!

File: examples/bert/bert_model.py [hunk 15]
```diff
@@ -0,0 +1,654 @@
...
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        activation = keras.activations.get(inner_activation)
+        initializer = keras.initializers.TruncatedNormal(
+            stddev=initializer_range
+        )
+        initializer = keras.initializers.get(initializer)
+
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/bert_model.py:541]: Yup, I've been testing on GPU for the most part. I think it just makes sure to do a one dimensional gather, which will run anywhere.

File: examples/bert/create_pretraining_data.py [hunk 16]
```diff
@@ -0,0 +1,524 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Create masked LM/next sentence masked_lm TF examples for BERT.
+
+This script will create TFRecord files containing BERT training examples with
+both word masking and next sentence prediction.
+
+This script will load the entire dataset into memory to setup the next sentence
+prediction task, so it is recommend to run this on shards of data at a time to
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/create_pretraining_data.py]: done

File: examples/bert/create_pretraining_data.py [hunk 17]
```diff
@@ -0,0 +1,524 @@
...
+        return text
+    elif isinstance(text, bytes):
+        return text.decode("utf-8", "ignore")
+    else:
+        raise ValueError("Unsupported string type: %s" % (type(text)))
+
+
+TrainingInstance = collections.namedtuple(
+    "TrainingInstance",
+    [
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/create_pretraining_data.py:126]: done

File: examples/bert/create_pretraining_data.py [hunk 18]
```diff
@@ -0,0 +1,524 @@
...
+    """Create `TrainingInstance`s from raw text."""
+    # Input file format:
+    # (1) One sentence per line. These should ideally be actual sentences, not
+    # entire paragraphs or arbitrary spans of text. (Because we use the
+    # sentence boundaries for the "next sentence prediction" task).
+    # (2) Blank lines between documents. Document boundaries are needed so
+    # that the "next sentence prediction" task doesn't span between documents.
+    dataset = tf.data.TextLineDataset(input_filenames)
+    dataset = dataset.map(
+        lambda x: tokenizer.tokenize(x).flat_values,
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/create_pretraining_data.py:217]: The bert tokenizer by default groups things by words (so there are two ragged dimensions for words and for subwords). flat_values removes this issue, but when we do our own tokenizer we can remove this behavior (or gate it behind a flag).

File: examples/bert/create_pretraining_data.py [hunk 19]
```diff
@@ -0,0 +1,524 @@
...
+                tokens.append("[SEP]")
+                segment_ids.append(0)
+
+                for token in tokens_b:
+                    tokens.append(token)
+                    segment_ids.append(1)
+                tokens.append("[SEP]")
+                segment_ids.append(1)
+
+                (
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/bert/create_pretraining_data.py:358]: This is just a black formatting choice, I tried rearranging into one line but blacks keeps reverting to this. I think we should just accept the autoformat.

File: examples/bert/create_pretraining_data.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/bert/create_pretraining_data.py:252]: SG

File: examples/bert/create_pretraining_data.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/bert/create_pretraining_data.py:459]: a follow up sg.

**Review by chenmoneygithub — APPROVED** [review]
Thanks!
